src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpUpgradeProject.cs

```diff
@@ -2,58 +2,60 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Test.Utilities;
 using Microsoft.VisualStudio.IntegrationTest.Utilities;
 using Roslyn.Test.Utilities;
 using Xunit;
-using Xunit.Abstractions;
-using ProjectUtils = Microsoft.VisualStudio.IntegrationTest.Utilities.Common.ProjectUtils;
 
 namespace Roslyn.VisualStudio.IntegrationTests.CSharp
 {
-    [Collection(nameof(SharedIntegrationHostFixture))]
-    public class CSharpUpgradeProject : AbstractUpdateProjectTest
+    [Trait(Traits.Feature, Traits.Features.CodeActionsUpgradeProject)]
+    public class CSharpUpgradeProject : AbstractUpgradeProjectTest
     {
-        public CSharpUpgradeProject(VisualStudioInstanceFactory instanceFactory)
-            : base(instanceFactory)
+        private async Task InvokeFixAsync(string version, CancellationToken cancellationToken)
         {
-        }
-
-        private void InvokeFix(string version = "latest")
-        {
-            VisualStudio.Editor.SetText(@$"
+            await TestServices.Editor.SetTextAsync(@$"
 #error version:{version}
-");
-            VisualStudio.Editor.Activate();
+", cancellationToken);
+            await TestServices.Editor.ActivateAsync(cancellationToken);
+
+            await TestServices.Editor.PlaceCaretAsync($"version:{version}", charsOffset: 0, cancellationToken);
 
-            VisualStudio.Editor.PlaceCaret($"version:{version}");
-            VisualStudio.Editor.InvokeCodeActionList();
-            VisualStudio.Editor.Verify.CodeAction($"Upgrade this project to C# language version '{version}'", applyFix: true);
+            // Suspend file change notification during code action application, since spurious file change notifications
+            // can cause silent failure to apply the code action if they occur within this block.
+            await using (var fileChangeRestorer = await TestServices.Shell.PauseFileChangesAsync(HangMitigatingCancellationToken))
```

Comment:
It might be useful to extract this helper to a core helper method on the EditorVerifiter perhaps?  e.g. `InvokeFixAsync(string name, applyFix:..., cancellationToken)` and re-use that here + CSharpUpdateProjectToAllowUnsafe.cs

Reply:
I'll keep it like this for now. The short term goal is eliminate this requirement (all cases where this appear are bugs that users will hit because they don't have the workaround).